Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -406,43 +438,66 @@ public async Task<string> WalkTree(string treeNodeKey)
+        /// <exception cref="OperationCanceledException">If the cancellation token was triggered.</exception>
+        /// <exception cref="Exception">If an unexpected exception was thrown.</exception>
+        /// <returns>The key of the next child to visit, or <c>null</c> if no match was found.</returns>
+        public async Task<string> VisitNode(string treeNodeKey, bool skipAction)
         {
             TreeNode treeNode = this.Schema.Tree[treeNodeKey];
 
-            // Do type-specific behavior.
-            switch (treeNode.Type)
+            if (!skipAction)
             {
-                case TreeNodeType.Leaf:
-                {
-                    await this.PerformLeafTypeBehavior(treeNode).ConfigureAwait(false);
-
-                    // Leaf type can't have ChildSelector so we return here.
-                    return null;
-                }
-                case TreeNodeType.Subroutine:
+                // Do type-specific behavior.
+                switch (treeNode.Type)
                 {
-                    // Exceptions are thrown here if the actions hit a timeout, were cancelled, or failed.
-                    await this.PerformSubroutineTypeBehavior(treeNode, treeNodeKey).ConfigureAwait(false);
-                    break;
-                }
-                case TreeNodeType.Action:
-                {
-                    // Exceptions are thrown here if the actions hit a timeout, were cancelled, or failed.
-                    await this.PerformActionTypeBehavior(treeNode, treeNodeKey).ConfigureAwait(false);
-                    break;
-                }
-                default:
-                {
-                    break;
+                    case TreeNodeType.Leaf:
+                        {
+                            await this.PerformLeafTypeBehavior(treeNode).ConfigureAwait(false);
+                            break;
+                        }
+                    case TreeNodeType.Subroutine:
+                        {
+                            // Exceptions are thrown here if the actions hit a timeout, were cancelled, or failed.
+                            await this.PerformSubroutineTypeBehavior(treeNode, treeNodeKey).ConfigureAwait(false);
+                            break;
+                        }
+                    case TreeNodeType.Action:
+                        {
+                            // Exceptions are thrown here if the actions hit a timeout, were cancelled, or failed.
+                            await this.PerformActionTypeBehavior(treeNode, treeNodeKey).ConfigureAwait(false);
+                            break;
+                        }
+                    default:
+                        {
+                            break;
+                        }
                 }
             }
 
+            if (treeNode.Type == TreeNodeType.Leaf)
+            {
+                // Leaf type can't have ChildSelector so we return here.
+                return null;
+            }
```

Comment:
Why did you move this?

Reply:
because regardless skipAction is true or false, we need the check below. So i moved the code to the bottom.

            if (treeNode.Type == TreeNodeType.Leaf)
            {
                // Leaf type can't have ChildSelector so we return here.
                return null;
            }

---
In reply to: [504218989](https://github.com/microsoft/Forge/pull/32#discussion_r504218989) [](ancestors = 504218989)